sdl_android/src/main/java/com/smartdevicelink/encoder/VirtualDisplayEncoder.java

```diff
@@ -0,0 +1,591 @@
+import android.os.Looper;
+import android.os.SystemClock;
+import android.util.DisplayMetrics;
+import android.util.Log;
+import android.view.Display;
+import android.view.MotionEvent;
+import android.view.Surface;
+import android.view.View;
+import android.view.Window;
+import android.view.WindowManager;
+
+import com.smartdevicelink.proxy.rpc.OnTouchEvent;
+import com.smartdevicelink.proxy.rpc.ScreenParams;
+import com.smartdevicelink.proxy.rpc.TouchCoord;
+import com.smartdevicelink.proxy.rpc.TouchEvent;
+import com.smartdevicelink.proxy.rpc.enums.TouchType;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.reflect.Constructor;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.FutureTask;
+
+
+public class VirtualDisplayEncoder {
+    private static final String TAG = "VirtualDisplayEncoder";
+    private static final int DENSITY = DisplayMetrics.DENSITY_HIGH;
+    private static final String VIDEO_MIME_TYPE = "video/avc";
+    private static int refresh_rate_ms = 100;
+    private static int video_width = 800;
+    private static int video_height = 480;
+    private static int frame_rate = 24;
+    private static int bitrate = 512000;
+    private static int interval = 5;
+    private DisplayManager mDisplayManager;
+    private volatile MediaCodec mVideoEncoder = null;
+    private volatile MediaCodec.BufferInfo mVideoBufferInfo = null;
+    volatile Surface inputSurface = null;
+    volatile VirtualDisplay virtualDisplay = null;
+    private volatile SdlPresentation presentation = null;
+    private Class<? extends SdlPresentation> presentationClass = null;
+    private VideoStreamWriterThread streamWriterThread = null;
+    private Context mContext;
+    private OutputStream sdlOutStream = null;
+    private Handler UI_handler = new Handler(Looper.getMainLooper());
+
+    public void init(Context context, OutputStream videoStream, Class<? extends SdlPresentation> presentationClass, ScreenParams screenParams) throws Exception {
+        if (android.os.Build.VERSION.SDK_INT < 21) {
+            Log.e(TAG, "API level of 21 required for VirtualDisplayEncoder");
+            throw new Exception("API level of 21 required");
+        }
+
+        mDisplayManager = (DisplayManager)context.getSystemService(Context.DISPLAY_SERVICE);
+
+        mContext = context;
+
+        if(screenParams.getImageResolution().getResolutionHeight() != null){
+            video_height = screenParams.getImageResolution().getResolutionHeight();
+        }
+
+        if(screenParams.getImageResolution().getResolutionWidth() != null){
+            video_width = screenParams.getImageResolution().getResolutionWidth();
+        }
+
+        sdlOutStream = videoStream;
+
+        this.presentationClass = presentationClass;
+
+        setupVideoStreamWriter();
+    }
+
+    public void setStreamParams(int frameRate, int bitrate, int interval, int refresh_rate){
```

Comment:
I'm leaning towards creating a local class that houses all the params available for video streaming instead of setting one by one.